pkg/dynamiclistener/server.go

```diff
@@ -85,6 +85,9 @@ func NewServer(ctx context.Context, listenConfigStorage ListenConfigStorage,
 }
 
 func (s *Server) updateIPs(savedIPs map[string]bool) map[string]bool {
+	s.Lock()
```

Comment:
this is bad because it will lock across multiple api calls. locking on network io is bad bc it introduce a huge bottleneck. can you reduce the scope of where you lock to just places where we iterate over s.certs and s.ips?

Reply:
Updated to just lock/unlock around s.certs loop. The LRU cache for s.ips contains its own mutex locking for safety.